File: node/examples/observer_peek_demo.ts
const timers = require("node:timers/promises");

type XprocModule = typeof import("../index");
type TransportOptions = import("../index").TransportOptions;
type Observer = import("../index").Observer;
type Consumer = import("../index").Consumer;

const xproc = require("../index.js") as XprocModule;

const delay = timers.setTimeout as (ms: number) => Promise<void>;

function cleanupShm(path: string): void {
  try {
    xproc.shmUnlink(path);
  } catch {
    // Best-effort cleanup only.
  }
}

async function waitForPeek(observer: Observer): Promise<Uint8Array> {
  while (true) {
    const payload = observer.peekCopy();
    if (payload !== null) {
      return payload;
    }
    await delay(1);
  }
}

async function waitForPoll(consumer: Consumer): Promise<Uint8Array> {
  while (true) {
    const payload = consumer.pollCopy();
    if (payload !== null) {
      return payload;
    }
    await delay(1);
  }
}

async function main(): Promise<void> {
  const shmPath = `/xproc_node_observer_demo_${process.pid}`;
  cleanupShm(shmPath);

  const createOptions: TransportOptions = {
    path: shmPath,
    shmSize: xproc.shmSizeForDataCapacity(16384n),
    channelType: "fixed",
    itemSize: 4,
    createIfMissing: true,
  };

  const attachOptions: TransportOptions = {
    ...createOptions,
    shmSize: xproc.XPROC_C_INFER_EXISTING_SHM_SIZE,
    createIfMissing: false,
  };

  const producer = new xproc.Producer(createOptions);
  const consumer = new xproc.Consumer(attachOptions);
  const observer = new xproc.Observer(attachOptions);

  try {
    const payload = Buffer.alloc(4);
    payload.writeInt32LE(0x42, 0);
    producer.sendFixedSized(payload);

    const peeked = await waitForPeek(observer);
    console.log(`observer sees: ${Buffer.from(peeked).readInt32LE(0)}`);

    const consumed = await waitForPoll(consumer);
    console.log(`consumer got: ${Buffer.from(consumed).readInt32LE(0)}, len=${consumed.length}`);

    const snapshot = observer.snapshot();
    console.log(
      `snapshot writePos=${snapshot.writePos} readPos=${snapshot.readPos} attachCount=${snapshot.attachCount}`,
    );
  } finally {
    observer.close();
    consumer.close();
    producer.close();
    cleanupShm(shmPath);
  }
}

void main().catch((error: unknown) => {
  const message = error instanceof Error ? error.stack ?? error.message : String(error);
  console.error(message);
  process.exit(1);
});

File: Python/examples/observer_peek_demo.py
from __future__ import annotations

import argparse
import os
import struct

from _common import cleanup_shm, load_xproc, wait_for_message


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--module-dir")
    args = parser.parse_args()

    xproc = load_xproc(args.module_dir)

    shm_path = f"/xproc_python_observer_demo_{os.getpid()}"
    cleanup_shm(xproc, shm_path)

    create_options = xproc.TransportOptions()
    create_options.path = shm_path
    create_options.shm_size = xproc.shm_size_for_data_capacity(16_384)
    create_options.channel_type = xproc.ChannelType.FIXED
    create_options.item_size = 4
    create_options.create_if_missing = True

    attach_options = xproc.TransportOptions()
    attach_options.path = shm_path
    attach_options.shm_size = xproc.INFER_EXISTING_SHM_SIZE
    attach_options.channel_type = xproc.ChannelType.FIXED
    attach_options.item_size = 4
    attach_options.create_if_missing = False

    producer = xproc.Producer(create_options)
    consumer = xproc.Consumer(attach_options)
    observer = xproc.Observer(attach_options)

    try:
        producer.send_fixed_sized(struct.pack("<i", 0x42))

        peeked = wait_for_message(observer.peek_copy, label="observer payload")
        print(f"observer sees: {struct.unpack('<i', peeked)[0]}")

        consumed = wait_for_message(consumer.poll_copy, label="consumer payload")
        print(f"consumer got: {struct.unpack('<i', consumed)[0]}, len={len(consumed)}")

        snapshot = observer.snapshot()
        print(
            "snapshot "
            f"write_pos={snapshot.write_pos} "
            f"read_pos={snapshot.read_pos} "
            f"attach_count={snapshot.attach_count}"
        )
    finally:
        observer.close()
        consumer.close()
        producer.close()
        cleanup_shm(xproc, shm_path)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
